source/common/signal_handler.cc

```diff
@@ -32,6 +34,7 @@ SignalHandler::SignalHandler(const std::function<void()>& signal_callback) {
 }
 
 SignalHandler::~SignalHandler() {
```

Comment:
If there was a real signal, does the process die and prevent this desturctor from being called? If so, makes perfect sense.

Reply:
So what this tries to fix is that without this change, the `DestructDoesNotFireHandler` test would fail, because here we would call the user-provided callback during destruction, even though no signal was raised. 

(This isn't a practical problem today when/where this happens because the callbacks themselves are resilient or called at a time where its effects no longer matter, but it is rather fragile)

So having said that, it sounds like you may have spotted a different tricky angle, but I'm not sure I follow; could you elaborate or rephrase to help me there?

Reply:
Sorry, I think I had a completely wrong view of how this works. I assumed there was some system magic that doesn't actually exist. I mainly missed the fact that you call `initiateShutdown()` in the destructor. 

It's definitely a good thing to skip the callback.

So to make sure I understand, is the following how it works in two possible scenarios? Hopefully there's no longer any interesting angle to speak of. If not, totally LGTM.

- Nighthawk process finishes without getting a signal
  - Destructor is called
    - The destructor calls `initiateShutdown()` which wakes up the shutdown thread via the pipe
    - The shutdown thread is smart enough to skip the callback (as of this PR)
    - Pipes get deleted
- Nighthawk process gets a signal like SIGTERM
  - Signal handler calls `initiateShutdown()` which wakes up the shutdown thread via the pipe
    - Callback gets called
    - Pipes get deleted
  - Destructor is called (?)
    - Destructor calls `initiateShutdown()`
      - But now the pipes are gone and nothing happens (?)

Reply:
That is all correct.  Revisiting this, this class could use a little love (code level docs, member renamings for clarity), it requires too much brainpower to follow as it stands. Thanks for persisting there, I'll file an issue, and your analysis here might then serve as a class level comment to guide others.

Reply:
(Filed https://github.com/envoyproxy/nighthawk/issues/603)